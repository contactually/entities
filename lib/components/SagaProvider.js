'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _getPrototypeOf = require('babel-runtime/core-js/object/get-prototype-of');

var _getPrototypeOf2 = _interopRequireDefault(_getPrototypeOf);

var _classCallCheck2 = require('babel-runtime/helpers/classCallCheck');

var _classCallCheck3 = _interopRequireDefault(_classCallCheck2);

var _createClass2 = require('babel-runtime/helpers/createClass');

var _createClass3 = _interopRequireDefault(_createClass2);

var _possibleConstructorReturn2 = require('babel-runtime/helpers/possibleConstructorReturn');

var _possibleConstructorReturn3 = _interopRequireDefault(_possibleConstructorReturn2);

var _inherits2 = require('babel-runtime/helpers/inherits');

var _inherits3 = _interopRequireDefault(_inherits2);

var _class, _temp;

var _propTypes = require('prop-types');

var _propTypes2 = _interopRequireDefault(_propTypes);

var _react = require('react');

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

var SagaProvider = (_temp = _class = function (_Component) {
  (0, _inherits3.default)(SagaProvider, _Component);

  function SagaProvider() {
    (0, _classCallCheck3.default)(this, SagaProvider);
    return (0, _possibleConstructorReturn3.default)(this, (SagaProvider.__proto__ || (0, _getPrototypeOf2.default)(SagaProvider)).apply(this, arguments));
  }

  (0, _createClass3.default)(SagaProvider, [{
    key: 'getChildContext',
    value: function getChildContext() {
      return {
        sagas: this.props.middleware
      };
    }
  }, {
    key: 'render',
    value: function render() {
      return _react.Children.only(this.props.children);
    }
  }]);
  return SagaProvider;
}(_react.Component), _class.propTypes = {
  children: _propTypes2.default.node,
  middleware: _propTypes2.default.func.isRequired
}, _class.childContextTypes = {
  sagas: _propTypes2.default.func.isRequired
}, _temp);
exports.default = SagaProvider;