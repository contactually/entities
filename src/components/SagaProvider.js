import React from 'react'
import PropTypes from 'prop-types'
import { Children, Component } from 'react'

class SagaProvider extends Component {
  static propTypes = {
    children: PropTypes.node,
    middleware: PropTypes.func.isRequired
  }

  static childContextTypes = {
    sagas: PropTypes.func.isRequired
  }

  getChildContext () {
    return {
      sagas: this.props.middleware
    }
  }
  render () {
    return Children.only(this.props.children)
  }
}

export default SagaProvider
