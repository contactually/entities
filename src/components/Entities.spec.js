import createSagaMiddleware from 'redux-saga'
import { applyMiddleware, createStore } from 'redux'
import React from 'react'
import { fromJS, Set, List, Map } from 'immutable'
import { Provider } from 'react-redux'
import SagaProvider from './SagaProvider'
import { mount } from 'enzyme'
import EntitiesModule from '../module'
import EntitiesConfig from '../Config'
import entities, {
  createMapState,
  createMapDispatch
} from './Entities'
import { omit } from 'lodash'

class MockApiClient {
  constructor () {
    this.get = jest.fn()
    this.put = jest.fn()
    this.post = jest.fn()
    this.delete = jest.fn()

    this.get.mockReturnValue(Promise.resolve({ data: {} }))
  }
}

const stubRequest = (state: {}) => {
  return fromJS({}).mergeDeep(fromJS(state))
}

const stubContainer = (state) => {
  return fromJS({
    activeRequests: Set(),
    latestRequest: null,
    scope: null,
    filters: Map(),
    ids: List(),
    meta: Map(),
    errors: Map()
  }).mergeDeep(fromJS(state))
}

const stubStore = (state: {entities?: {}, containers?: {}, requests?: {}}) => {
  return fromJS({
    entities: {
      entities: {},
      containers: {},
      requests: {}
    }
  }).mergeDeep({entities: fromJS(state)})
}

describe('components.Entities', () => {
  const containers = {
    contactResource: {type: 'contact'},
    bucketResource: {type: 'bucket', autoload: false}
  }
  const containerKeys = Object.keys(containers)
  const initialState = stubStore({})
  const initialProps = {
    contactResource: {
      containerId: 'contacts-container'
    },
    bucketResource: {
      containerId: 'buckets-container'
    }
  }

  describe('entities', () => {
    let component
    beforeEach(() => {
      EntitiesConfig.configure({
        apiClient: new MockApiClient()
      })
      const sagaMiddleware = createSagaMiddleware()
      const store = createStore(
        EntitiesModule.rootReducer,
        stubStore({}),
        applyMiddleware(sagaMiddleware)
      )
      sagaMiddleware.run(EntitiesModule.rootSaga)
      const Test = () => <div />
      const WrappedTest = entities(containers)(Test)
      component = mount(
        <SagaProvider middleware={sagaMiddleware}>
          <Provider store={store}>
            <WrappedTest />
          </Provider>
        </SagaProvider>
      )
    })

    it('renders', () => {
      expect(component.length).toEqual(1)
    })

    it('has correct props for contactResource', () => {
      const contactResourceProps = component.find('Test').at(0).prop('contactResource')
      expect(omit(contactResourceProps, ['containerId'])).toMatchSnapshot()
    })

    it('has correct props for bucketResource', () => {
      const bucketResourceProps = component.find('Test').at(0).prop('bucketResource')
      expect(omit(bucketResourceProps, ['containerId'])).toMatchSnapshot()
    })
  })

  describe('createMapState', () => {
    it('returns a memoized function', () => {
      const mapState = createMapState(containers, containerKeys)
      const result = mapState(initialState, initialProps)
      expect(result).toEqual(expect.any(Function))
    })

    it('gets state', () => {
      const mapState = createMapState(containers, containerKeys)
      const resultFn = mapState(initialState, initialProps)
      const globalState = stubStore({
        containers: {
          'contacts-container': stubContainer({}),
          'buckets-container': stubContainer({})
        }
      })
      const props = {}
      const result = resultFn(globalState, props)
      expect(fromJS(result).toJS()).toMatchSnapshot()
    })

    it('sets isLoading when a container has activeRequests', () => {
      const mapState = createMapState(containers, containerKeys)
      const resultFn = mapState(initialState, initialProps)
      const globalState = stubStore({
        containers: {
          'contacts-container': stubContainer({
            activeRequests: Set([stubRequest({})])
          }),
          'buckets-container': stubContainer({})
        }
      })
      const props = {}
      const result = resultFn(globalState, props)
      expect(result.contactResource.isLoading).toEqual(true)
      expect(result.bucketResource.isLoading).toEqual(false)
      expect(result).toMatchSnapshot()
    })

    it('sets isFetched when a container has latestRequest', () => {
      const mapState = createMapState(containers, containerKeys)
      const resultFn = mapState(initialState, initialProps)
      const globalState = stubStore({
        containers: {
          'contacts-container': stubContainer({
            latestRequest: stubRequest({})
          }),
          'buckets-container': stubContainer({})
        }
      })
      const props = {}
      const result = resultFn(globalState, props)
      expect(result.contactResource.isFetched).toEqual(true)
      expect(result.bucketResource.isFetched).toEqual(false)
      expect(result).toMatchSnapshot()
    })
  })

  describe('createMapDispatch', () => {
    const initialDispatch = jest.fn()

    it('returns a memoized function', () => {
      const mapDispatch = createMapDispatch(containers, containerKeys)
      const result = mapDispatch(initialDispatch, initialProps)
      expect(result).toEqual(expect.any(Function))
    })

    describe('binding actions', () => {
      let actions
      beforeEach(() => {
        const mapDispatch = createMapDispatch(containers, containerKeys)
        const resultFn = mapDispatch(initialDispatch, initialProps)
        actions = resultFn()
      })

      it('fetchAll', () => {
        const result = actions.contactResource.fetchAll()
        expect(initialDispatch).toBeCalled()
      })

      it('initialize', () => {
        const result = actions.contactResource.initialize()
        expect(initialDispatch).toBeCalled()
      })

      it('performAction', () => {
        const action = 'fetch'
        const options = { itemId: 'contact_123' }
        const result = actions.contactResource.performAction(action, options)
        expect(initialDispatch).toBeCalled()
        expect(initialDispatch.mock.calls[initialDispatch.mock.calls.length - 1]).toMatchSnapshot()
      })

      it('updateFilters', () => {
        const result = actions.contactResource.updateFilters()
        expect(initialDispatch).toBeCalled()
      })

      it('resetContainer', () => {
        const result = actions.contactResource.resetContainer()
        expect(initialDispatch.mock.calls[initialDispatch.mock.calls.length - 1]).toMatchSnapshot()
      })
    })
  })
})
