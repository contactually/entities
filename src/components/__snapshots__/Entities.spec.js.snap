// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`components.Entities createMapDispatch binding actions performAction 1`] = `
Array [
  Object {
    "error": undefined,
    "payload": Object {
      "action": "fetch",
      "containerId": "contacts-container",
      "itemId": "contact_123",
      "type": "contact",
    },
    "type": "S/entities/PERFORM_ACTION/TRIGGER",
  },
]
`;

exports[`components.Entities createMapState gets state 1`] = `
Object {
  "bucketResource": Object {
    "errors": Object {},
    "filters": Object {},
    "isFetched": false,
    "isLoading": false,
    "item": undefined,
    "items": Array [],
    "meta": Object {},
    "missingIds": Array [],
    "total": 0,
  },
  "contactResource": Object {
    "errors": Object {},
    "filters": Object {},
    "isFetched": false,
    "isLoading": false,
    "item": undefined,
    "items": Array [],
    "meta": Object {},
    "missingIds": Array [],
    "total": 0,
  },
}
`;

exports[`components.Entities createMapState sets isFetched when a container has latestRequest 1`] = `
Object {
  "bucketResource": Object {
    "errors": Immutable.Map {},
    "filters": Immutable.Map {},
    "isFetched": false,
    "isLoading": false,
    "item": undefined,
    "items": Immutable.List [],
    "meta": Immutable.Map {},
    "missingIds": Immutable.Set [],
    "total": 0,
  },
  "contactResource": Object {
    "errors": Immutable.Map {},
    "filters": Immutable.Map {},
    "isFetched": true,
    "isLoading": false,
    "item": undefined,
    "items": Immutable.List [],
    "meta": Immutable.Map {},
    "missingIds": Immutable.Set [],
    "total": 0,
  },
}
`;

exports[`components.Entities createMapState sets isLoading when a container has activeRequests 1`] = `
Object {
  "bucketResource": Object {
    "errors": Immutable.Map {},
    "filters": Immutable.Map {},
    "isFetched": false,
    "isLoading": false,
    "item": undefined,
    "items": Immutable.List [],
    "meta": Immutable.Map {},
    "missingIds": Immutable.Set [],
    "total": 0,
  },
  "contactResource": Object {
    "errors": Immutable.Map {},
    "filters": Immutable.Map {},
    "isFetched": false,
    "isLoading": true,
    "item": undefined,
    "items": Immutable.List [],
    "meta": Immutable.Map {},
    "missingIds": Immutable.Set [],
    "total": 0,
  },
}
`;

exports[`components.Entities entities has correct props for bucketResource 1`] = `
Object {
  "errors": Immutable.List [],
  "fetchAll": [Function],
  "filters": Immutable.Map {},
  "initialize": [Function],
  "isFetched": false,
  "isLoading": false,
  "item": undefined,
  "items": Immutable.List [],
  "meta": Immutable.Map {},
  "missingIds": Immutable.Set [],
  "performAction": [Function],
  "total": 0,
  "updateCollection": [Function],
  "updateFilters": [Function],
}
`;

exports[`components.Entities entities has correct props for contactResource 1`] = `
Object {
  "errors": Immutable.List [],
  "fetchAll": [Function],
  "filters": Immutable.Map {},
  "initialize": [Function],
  "isFetched": false,
  "isLoading": false,
  "item": undefined,
  "items": Immutable.List [],
  "meta": Immutable.Map {},
  "missingIds": Immutable.Set [],
  "performAction": [Function],
  "total": 0,
  "updateCollection": [Function],
  "updateFilters": [Function],
}
`;
